src/core/eggmath.rkt

```diff
@@ -0,0 +1,213 @@
+#lang racket
+
+(require ffi/unsafe
+         ffi/unsafe/define)
+(require racket/runtime-path)
+
+(require "../debug.rkt" "../common.rkt")
+(require "../syntax/rules.rkt")
+(require "../syntax/herbie-to-egg-pattern.rkt")
+
+(require egg-herbie)
+
+(module+ test (require rackunit))
+
+(provide egraph-run egraph-add-exprs egraph-run-iter egraph-get-simplest egg-expr->expr egg-add-exn? make-ffi-rules egraph-get-cost egraph-get-size)
+
+
+(struct egraph-data (egraph-pointer egg->herbie-dict herbie->egg-dict))
+
+
+(define (egraph-get-size egraph-data)
+  (egraph_get_size (egraph-data-egraph-pointer egraph-data)))
+
+(define (egraph-get-cost egraph-data node-id)
+  (egraph_get_cost (egraph-data-egraph-pointer egraph-data) node-id))
+
+(define (egraph-get-simplest egraph-data node-id)
+  (egraph_get_simplest (egraph-data-egraph-pointer egraph-data) node-id))
+
+(define (make-ffi-rules rules)
+  (for/list [(rule rules)]
+      (make-FFIRule (symbol->string (rule-name rule))
+                    (herbie-pattern->rust-pattern (rule-input rule))
+                    (herbie-pattern->rust-pattern (rule-output rule)))))
+
+(define (egraph-run-iter egraph-data node-limit ffi-rules precompute?)
+  (egraph_run_iter (egraph-data-egraph-pointer egraph-data) node-limit ffi-rules precompute?))
+
+
+(define (egraph-run-rules egraph-data node-limit rules precompute?)
+  (define ffi-rules (make-ffi-rules rules))
+  (define old-cnt 0)
+  (for/and ([iter (in-naturals 0)])
+    (egraph_run_iter (egraph-data-egraph-pointer egraph-data) node-limit ffi-rules precompute?)
+    (define cnt (egraph_get_size (egraph-data-egraph-pointer egraph-data)))
+    (define is_stop (or (>= cnt node-limit) (<= cnt old-cnt)))
```

Comment:
I don't really like using mutation for `old-cnt` here, especially since it's so easy to avoid, and since it forces you to define `is_stop` and then using it later.